sklearn/tree/_criterion.pyx

```diff
@@ -29,52 +31,31 @@ from ._utils cimport WeightedMedianCalculator
 # EPSILON is used in the Poisson criterion
 cdef double EPSILON = 10 * np.finfo('double').eps
 
-cdef class Criterion:
-    """Interface for impurity criteria.
+cdef class BaseCriterion:
+    """This is an abstract interface for criterion. For example, a tree model could
+    be either supervisedly, or unsupervisedly computing impurity on samples of
+    covariates, or labels, or both. Although scikit-learn currently only contains
+    supervised tree methods, this class enables 3rd party packages to leverage
+    scikit-learn's Cython code for criteria.
+
+    The downstream classes _must_ implement methods to compute the impurity
+    in current node and in children nodes.
```

Comment:
PEP257 for better integration with API doc tools like sphinx and IDE user interfaces (e.g. code navigation and completion). First line should be a short one-liner summary.

Let me also suggest more details about the lack of API compat guarantees:

```suggestion
cdef class BaseCriterion:
    """Abstract base class for all criterion implementations.
    
    For example, a tree model could be either supervisedly, or unsupervisedly
    computing impurity on samples of covariates, or labels, or both.
     
    Although scikit-learn currently only contains supervised tree methods,
    this class enables 3rd party packages to leverage scikit-learn's Cython
    code for criteria without too much duplication (but without any backward
    compatibility guarantees between scikit-learn releases).

    The downstream classes _must_ implement methods to compute the impurity
    in current node and in children nodes.
```

Reply:
Adding to the upstream PR for `Criterion` #24678